mono/metadata/threads.c

```diff
@@ -3921,84 +3938,31 @@ void mono_thread_suspend_all_other_threads (void)
 	 */
 	g_assert (shutting_down);
 
+	mono_threads_lock ();
+	mono_g_hash_table_foreach (threads, suspend_thread, NULL);
+	mono_threads_unlock ();
+
 	/*
-	 * We make multiple calls to WaitForMultipleObjects since:
-	 * - we can only wait for MONO_W32HANDLE_MAXIMUM_WAIT_OBJECTS threads
-	 * - some threads could exit without becoming suspended
+	 * Cooperative suspend requires the threads to be scheduled and enter
+	 * suspend themselves. If we won't wait there's a race against any runtime
+	 * cleanup that's done via mono_runtime_quit_internal() and may reference data
+	 * structures that were already tore down. If the thread is stuck in a prolonged
+	 * native call there's not much that can be done to interrupt it. Best we can do
+	 * is give up after certain amount of time and then continue with shutdown.
 	 */
-	finished = FALSE;
-	while (!finished) {
-		/*
-		 * Make a copy of the hashtable since we can't do anything with
-		 * threads while threads_mutex is held.
-		 */
-		wait->num = 0;
-		/*We must zero all InternalThread pointers to avoid making the GC unhappy.*/
-		memset (wait->threads, 0, sizeof (wait->threads));
+	start_time = mono_msec_ticks ();
+	while (still_suspending && timeout >= 0) {
+		still_suspending = FALSE;
 		mono_threads_lock ();
-		mono_g_hash_table_foreach (threads, collect_threads_for_suspend, wait);
+		mono_g_hash_table_foreach (threads, any_unsuspended_threads, &still_suspending);
 		mono_threads_unlock ();
+		mono_thread_info_sleep (50, NULL);
+		timeout -= mono_msec_ticks() - start_time;
+	}
```

Comment:
I'm not sure, but I think this counts the time quadratically? You're subtracting time since the start each iteration. I think it'd make more sense to store the end time and compare to that.

Reply:
Correct, I forgot to stage a line that was reassigning `start_time = mono_msec_ticks ();`. I'll push that in a moment.

Reply:
It's a bit nit-picky, but I feel like this wait loop is way more complicated than it needs to be. I'm hung up on there being two calls to `mono_msec_ticks` in the loop, which shouldn't make a practical difference but in theory there could be time passing between them. You could just assign start_time once and check `mono_msec_ticks() - start_time < 1000`.

Reply:
Also, if I'm reading this right, we always wait 50 ms, even if `still_suspending` was set to TRUE on this iteration. The final iteration will wait, but it won't check whether the threads finished suspending.

I think I would add a helper function for checking for suspended threads and write the loop something like this:

```c
start_time = mono_msec_ticks ();
while ((still_suspending = is_still_suspending()) && mono_msec_ticks () - start_time < 1000)
{
    mono_thread_info_sleep (50, NULL);
}

if (still_suspending) emit warning and skip shutdown
```

Reply:
I've based the original logic on some other timeouts implemented in the same file. I've changed the implementation to match your suggestion.